### DISCUSSION TITLE
Prevent Deletion of DecoratorBlockNode With Backspace Command

### DISCUSSION THREAD

**Discussion by lsbyerley** [facebook/lexical#5195] [General]
I have a created a custom DecoratorBlockNode used for displaying media (image, video, meta info) in the editor. There is a button inside this node used for deleting the node. I would like to prevent the user from deleting this node when typing in the editor (ie. hitting the backspace command).

Current approach to this is below. The caveat is, it seems to work fine when the backspace is pressed while on an empty line right before the node. It does not work when the backspace is pressed on a line with text. Does anyone know why that might be?

```
const onBackspaceCommand = useCallback(() => {
  const selection = $getSelection();
  const nodes = selection.getNodes();
  const node = nodes[0];
  const prevNode = node.getPreviousSibling();

  // Prevent custom dedcorator node deletion with backspace command
  if ($isCustomDecoratorNode(node) || $isCustomDecoratorNode(prevNode)) {
    return true;
  }

  return false;
}, []);

useEffect(() => {
  return mergeRegister(
    editor.registerCommand(KEY_BACKSPACE_COMMAND, onBackspaceCommand, COMMAND_PRIORITY_LOW)
  );
}, []);
```

![Screenshot 2023-11-02 at 9 19 47 AM](https://github.com/facebook/lexical/assets/3066258/fcf07817-a979-44c7-9f34-688087b510d7)

![Screenshot 2023-11-02 at 9 20 44 AM](https://github.com/facebook/lexical/assets/3066258/c64e625d-7560-4251-ba54-f4094eec3074)

**Comment by 2wheeh** [comment]
My guess:

```ts
  // Prevent custom dedcorator node deletion with backspace command
  if ($isCustomDecoratorNode(node) || $isCustomDecoratorNode(prevNode)) {
    return true;
  }
```
This checks 
1. sibling of selected node is `CustomDecoraterNode`
2. selected node itself is `CustomDecoraterNode`

When there is not any text content, selected node is paragraph node and it is case 1 as `prevNode`  (or it's childNode ?) is `CustomDecoratorNode` :
<img width="492" alt="image" src="https://github.com/facebook/lexical/assets/40269597/ec1a7935-fc4d-49a1-8b33-1e4c2e656fc8">

But when it has a text content, selected node is a text node. so It's not a `CustomDecoratorNode`.
and this text node has no siblings (`prevNode` is probably `null`) :
<img width="465" alt="image" src="https://github.com/facebook/lexical/assets/40269597/9e534394-eca0-4c0d-bc90-0453f7f87046">

To achieve what you want, It seems you should check uncle (or it's child) of selected node as well.

**Reply by 2wheeh** [reply]
So I guess it should look more like this  ...?

```ts
const onBackspaceCommand = useCallback(() => {
    const selection = $getSelection();

    // Prevent custom decorator node deletion with backspace command
    if ($isRangeSelection(selection) && selection.isCollapsed()) {
      if (selection.anchor.offset === 0) {
        const element =
          selection.anchor.type === 'element'
            ? selection.anchor.getNode()
            : selection.anchor.getNode().getParentOrThrow();

        const prevNode = element.getPreviousSibling();
        if ($isCustomDecoratorNode(prevNode)) {
          return true;
        }
      } else {
        const curNode = selection.anchor.getNode();
        if ($isCustomDecoratorNode(curNode)) {
          return true;
        }
      }
    }

    return false;
  }, []);

  useEffect(() => {
    return mergeRegister(
      editor.registerCommand(KEY_BACKSPACE_COMMAND, onBackspaceCommand, COMMAND_PRIORITY_LOW));
  }, [editor, onBackspaceCommand]);
  ```

**Reply by lsbyerley** [reply]
@2wheeh Thank you for the detailed reply. I think your logic does a good job of encapsulating everything needed to achieve this. Whats odd is, after testing your code, the CustomDecoratorNode still gets deleted even when returning true inside the node check:

```
if ($isCustomDecoratorNode(prevNode)) {
 // execution gets into here and returns true, but node is still deleted
 return true;
}
```

Do you know why that could be?

**Reply by 2wheeh** [reply]
> Do you know why that could be?

It may be core `KEY_BACKSPACE_COMMAND` listener deleting `CustomDecoratorNode` has a higher priority of yours (or just enrolled earlier than yours). 
try `useLayoutEffect` or higher priority (like `PRIORITY_CRITICAL`) or both.
Not sure this is proper usage of it 😅 ... I hope this will work !

**Reply by lsbyerley** [reply]
So it does appear `DELETE_CHARACTER_COMMAND` is also firing.. I have since added it to the command list. However, I am still noticing the same behavior. There must be something else firing as well but I'm not sure which command it could be. It is not `DELETE_LINE_COMMAND` or `DELETE_WORD_COMMAND`

I've also bumped up priority to `COMMAND_PRIORITY_CRITICAL`

**Reply by 2wheeh** [reply]
Have you tried this:
```ts
 useLayoutEffect(() => {
    return mergeRegister(
      editor.registerCommand(
        DELETE_CHARACTER_COMMAND, // <-- instead of KEY_BACKSPACE_COMMAND
        onBackspaceCommand,
        COMMAND_PRIORITY_CRITICAL,
      ),
    );
  }, [editor, onBackspaceCommand]);
  ```

**Reply by lsbyerley** [reply]
@2wheeh Just gave this a try and unfortunately yields the same result :(

```
useLayoutEffect(() => {
    return mergeRegister(
      editor.registerCommand(KEY_BACKSPACE_COMMAND, onBackspaceCommand, COMMAND_PRIORITY_CRITICAL),
      editor.registerCommand(DELETE_CHARACTER_COMMAND, onBackspaceCommand, COMMAND_PRIORITY_CRITICAL)
    )
  }, [editor, onBackspaceCommand]);
```

**Reply by lsbyerley** [reply]
What's also frustrating is the debugger window only shows those two commands as well..

```

 commands:
  └ { type: CLICK_COMMAND, payload: PointerEvent }
  └ { type: KEY_BACKSPACE_COMMAND, payload: KeyboardEvent }
  └ { type: DELETE_CHARACTER_COMMAND, payload: true }
  └ { type: CAN_UNDO_COMMAND, payload: true }
```

**Reply by lsbyerley** [reply]
On second thought, it appears upping the priority level and keeping the registered command in a useEffect does work! I'll close this out

```
const onBackspaceCommand = useCallback(() => {
  const selection = $getSelection();

  if ($isRangeSelection(selection) && selection.isCollapsed()) {
    if (selection.anchor.offset === 0) {
      const element = selection.anchor.type === 'element' ? selection.anchor.getNode() : selection.anchor.getNode().getParentOrThrow();

      const prevNode = element.getPreviousSibling();
      if ($isDecoratorNode(prevNode)) {
        return true;
      }
    } else {
      const curNode = selection.anchor.getNode();
      if ($isDecoratorNode(curNode)) {
        return true;
      }
    }
  }

  return false;
}, []);

useEffect(() => {
  return mergeRegister(
    editor.registerCommand(KEY_BACKSPACE_COMMAND, onBackspaceCommand, COMMAND_PRIORITY_CRITICAL),
    editor.registerCommand(DELETE_CHARACTER_COMMAND, onBackspaceCommand, COMMAND_PRIORITY_CRITICAL)
  );
}, []);
```
